tracer/src/Datadog.Trace/DataStreamsMonitoring/Utils/BinaryPrimitivesHelper.cs

```diff
@@ -0,0 +1,76 @@
+﻿// <copyright file="BinaryPrimitivesHelper.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+
+namespace Datadog.Trace.DataStreamsMonitoring.Utils;
+
+internal static class BinaryPrimitivesHelper
+{
+    public static void WriteUInt64LittleEndian(byte[] bytes, ulong value)
+    {
+#if NETCOREAPP3_1_OR_GREATER
+        System.Buffers.Binary.BinaryPrimitives.WriteUInt64LittleEndian(bytes, value);
+#else
+        // write the bytes one at a time
+        if (BitConverter.IsLittleEndian)
+        {
+            bytes[0] = (byte)(value & 0xFF);
+            bytes[1] = (byte)((value >> 8) & 0xFF);
+            bytes[2] = (byte)((value >> 16) & 0xFF);
+            bytes[3] = (byte)((value >> 24) & 0xFF);
+            bytes[4] = (byte)((value >> 32) & 0xFF);
+            bytes[5] = (byte)((value >> 40) & 0xFF);
+            bytes[6] = (byte)((value >> 48) & 0xFF);
+            bytes[7] = (byte)((value >> 56) & 0xFF);
+        }
+        else
+        {
+            bytes[8] = (byte)(value & 0xFF);
```

Comment:
I'm a little skeptical of this else branch. If you do `& 0xFF` on a `ulong`, I thought that would always get the bottom byte of the `ulong`. If that's the case, then we should always be assigning that result to `bytes[0]` since we are storing the value as little endian. I suggested writing some tests below as a way to validate this

Reply:
er, yeah, I'm sceptical too, it clearly gives an index out of range exception given I'm accessing `bytes[8]` 🤦‍♂️ Wasn't caught because these branches are entirely untested, as we don't have bigendian machines 🤦‍♂️ 

If it makes you feel any more comfortable, this algorithm is the one Jon Skeet uses in his "classic" Misc Utils lib: https://github.com/timdetering/JonSkeet.MiscUtil/blob/e2413fa33958a56afde0926f1db14d01aa04d3c7/src/MiscUtil/Conversion/BigEndianBitConverter.cs#L41-L46

Honestly, I'm not entirely sure how to test this... I could pull the "big endian" branch out to test it explicitly, but then, I need the machine to actually store the ulong in a big endian way, which I can't. Unless I mock it. In which case I'm no longer testing against anything reliable. (FWIW, Jon Skeet only tests the little-endian converters too 😅)

So I figure I _either_ leave it as is (after fixing the indexing issue) or explicitly throw in this branch. Neither is particularly appealing, but also, basically nothing uses bigendian these days AFAIK?

Reply:
> I thought that would always get the bottom byte of the ulong

Just talking this through again to make sure. This statement is true, you do get the "bottom" byte of the `ulong`, but the bottom byte is the Most Significant Bits, so it needs to go in `bytes[7]`

So, considering an int (for simplicity)

```
// [0x11 0x22 0x33 0x44] little endian 
// [0x44 0x33 0x22 0x11] big endian 
int x = 287,454,020; 

// [0x11 0x22 0x33 0x44] & [0xFF] = 0x44 little endian 
// [0x44 0x33 0x22 0x11] & [0xFF] = 0x11 big endian 
int y = x & 0xFF;
```

Given we need to store as  [0x11 0x22 0x33 0x44], we need to reverse the order of the copying.

I hope that's right 😅

Reply:
Ah I see what you're saying, thanks for walking through that. Ok I'm content with the current behavior 😄